### PR TITLE
OpenID Connect Discovery Support

### PR DESCRIPTION
**Pull request by cloudcogsio**
Initial Commit
Tests and Documentation to follow.

### CODE REVIEW

**Comment by cloudcogsio** [comment]
Proposed

1. New ```AbstractOIDCProvider``` that extends the base ```AbstractProvider``` but provides mechanisms for OIDC.

2. Custom Providers wishing to utilize OIDC should extend the ```AbstractOIDCProvider``` class.

3. An Interface which should be implemented and passed to the Provider config that will handle caching of public keys (JWKs). A simple file system implementation is provided. ```League\OAuth2\Client\Provider\OpenIDConnect\PublicKeyCache\File```

4. Required configs are [clientId, clientSecret, well_known_endpoint, publickey_cache_provider]

- **well_known_endpoint** - The URL of the well-known endpoint service of the provider
- **publickey_cache_provider** - A concrete implementation of ```League\OAuth2\Client\Provider\OpenIDConnect\PublicKeyCacheInterface```

```PHP

use League\OAuth2\Client\Provider\OpenIDConnect\AbstractOIDCProvider;

class MyProvider extends AbstractOIDCProvider 
{
    /**
     * {@inheritDoc}
     * @see \League\OAuth2\Client\Provider\AbstractProvider::getDefaultScopes()
     */
    protected function getDefaultScopes()
    {
        // TODO Auto-generated method stub
        
    }

    /**
     * {@inheritDoc}
     * @see \League\OAuth2\Client\Provider\AbstractProvider::checkResponse()
     */
    protected function checkResponse(\Psr\Http\Message\ResponseInterface $response, $data)
    {
        // TODO Auto-generated method stub
        
    }

    /**
     * {@inheritDoc}
     * @see \League\OAuth2\Client\Provider\AbstractProvider::createResourceOwner()
     */
    protected function createResourceOwner(array $response, \League\OAuth2\Client\Token\AccessToken $token)
    {
        // TODO Auto-generated method stub
        
    }

    
}

$Provider = new MyProvider([
    AbstractOIDCProvider::OPTION_WELL_KNOWN_URL => 'https://my.auth.server/.well-known/openid-configuration',
    AbstractOIDCProvider::OPTION_PUBLICKEY_CACHE_PROVIDER => new \League\OAuth2\Client\Provider\OpenIDConnect\PublicKeyCache\File('my.auth.server.keys'),
    'clientId' => '{CLIENT-ID}',
    'clientSecret' => '{CLIENT-SECRET}'
]);

// Dump the auto discovered data
print_r($Provider->Discovery());

// Get the base auth URL
print $Provider->Discovery()->getAuthorizationEndpoint();


```

**Comment by ruben0909** [comment]
Thank you for this :)

**Comment by cloudcogsio** [comment]
> Thank you for this :)

You're welcome.

Decided to make it available as a separate package.
https://github.com/cloudcogsio/oauth2-openid-connect-discovery

May close this PR or leave pending for 3.x consideration.

File: composer.json [hunk 1]
```diff
@@ -1,68 +1,66 @@
 {
-    "name": "league/oauth2-client",
```
Comment (ramsey) [inline composer.json:2]: Please don't modify the indentation of the composer.json file. Thanks!

Comment (cloudcogsio) [reply composer.json]: > 

IDE issues.
Should be fixed now.

**Comment by pierrocknroll** [comment]
Is there any plan merge this PR ? Thanks

**Pull request closed without merging**